### PR TITLE
Update Ruff and enable F821 in stubs

### PR DESCRIPTION
**Pull request by Avasam**
Thanks to @AlexWaygood 's work on https://github.com/astral-sh/ruff/issues/3011 , we no longer need to ignore `F821` in Ruff configs, whilst still preferring to avoid forward-references.

I've left in `F821` from Flake8 as per my in-code comment: typeshed being a testing ground for F821, it's great to catch changes to the monkeypatching of `F821`

### CODE REVIEW

**Review by AlexWaygood — APPROVED** [review]
Thanks! I actually wouldn't be opposed to ignoring flake8's version of F821 now, but we can leave that discussion for another time/it's a separate thing that should probably be considered separately

**Comment by Avasam** [comment]
Me neither, but I figured some of the reasoning about testing / feedback loop in https://github.com/python/typeshed/pull/11734#issuecomment-2043455964 could apply here as well. Although Flake8-pyi's unit tests *should* essentially cover changes and regressions.

Happy to make a follow-up PR to see what others think (mainly from flake8-pyi maintainers).

File: stdlib/typing.pyi [hunk 1]
```diff
@@ -132,9 +132,6 @@ if sys.version_info >= (3, 12):
 ContextManager = AbstractContextManager
 AsyncContextManager = AbstractAsyncContextManager
 
-# This itself is only available during type checking
-def type_check_only(func_or_cls: _F) -> _F: ...
```
Comment (JelleZijlstra) [inline stdlib/typing.pyi:136]: Why were these moved? This is legal in stubs, even if it might be better style to move the function after the definition of `_F`.

Comment (AlexWaygood) [reply stdlib/typing.pyi:136]: (If it does get added to a spec somewhere or we think there _is_ a good reason to support _arbitrary_ forward references in stubs, I'm personally very open to adding support for it at Ruff.)

Comment (JelleZijlstra) [reply stdlib/typing.pyi:136]: That makes sense in general; the second change in this PR (`connect = Connection` before `Connection` is defined) is definitely some very odd code.

I'm not sure why Ruff has an issue with the definition of `def type_check_only`, though. `_F` is used only in an annotation; why does the definition of `type_check_only` need to be after the definition of `_F`?

Comment (AlexWaygood) [reply stdlib/typing.pyi:136]: The issue is that the TypeVar `_F` is bound to `Callable` before `Callable` is defined

Comment (JelleZijlstra) [reply stdlib/typing.pyi:136]: Ah, I guess that's the kind of case we only encounter in typeshed :)

Comment (AlexWaygood) [reply stdlib/typing.pyi:136]: Yup! 😄
